include/swift/AST/SourceFile.h

```diff
@@ -125,11 +125,28 @@ class SourceFile final : public FileUnit {
   /// been validated.
   llvm::SetVector<ValueDecl *> UnvalidatedDeclsWithOpaqueReturnTypes;
 
+  /// The list of top-level declarations in the source file.
+  std::vector<Decl *> Decls;
+
   friend ASTContext;
   friend Impl;
+
 public:
-  /// The list of top-level declarations in the source file.
-  std::vector<Decl*> Decls;
+  /// Appends the given declaration to the end of the top-level decls list.
+  void addTopLevelDecl(Decl *d) {
```

Comment:
Eventually this should set them all in one shot so that anything depending on getTopLevelDecls() is immutable.

Reply:
LLDB is going to be the last bastion of mutations here...